dG Dependecies: Mesh Modifiers and Subdivision

int-ptr-ptr opened the pull request:
# Description

*Please describe the changes/features in this pull request.*

This PR adds a `mesh-modifiers` tree to the specfem parameters, as well as its first entry: `subdivisions`, which allow for modifying `specfem::mesh::mesh` by subdividing elements corresponding to a specific material. Upon the introduction of LCIs, a `mesh-modifiers.coupling` entry will allow the user to specify how to resolve interfaces (strongly coupled / loosely coupled, what flux scheme, etc.), but for now, if adjacent materials have nonconforming subdivisions, the code should error out.

Going forward, `mesh-modifiers` should add a new step to the `execute()` pipeline, where, after reading in a `specfem::mesh:mesh` instance, a `modifiers.apply(&mesh)` call would augment the mesh before being passed into the constructor of `specfem::compute::assembly`.

# Issue Number

*If there is an issue created for these changes, link it here*

# Checklist

*Please make sure to check developer documentation on specfem docs. *

- [x] I ran the code through pre-commit to check style
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [ ] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

buildbot-princeton commented:
Can one of the admins verify this patch?

buildbot-princeton commented:
Can one of the admins verify this patch?

buildbot-princeton commented:
Can one of the admins verify this patch?

Rohit-Kakodkar commented:
ok to test

lsawade commented:
Is this working, or work in progress? If it is working could you add an example to the examples directory?

lsawade reviewed (approved):
Ooops, reviews this a bit ago but forgot to submit the review...

LGTM

Rohit-Kakodkar commented on src/mesh/modifiers/modifiers.cpp:
```diff
@@ -0,0 +1,52 @@
+}
+
+template class specfem::mesh::modifiers<specfem::dimension::type::dim2>;
+template class specfem::mesh::modifiers<specfem::dimension::type::dim3>;
```
Do you specialize apply function for 3D mesh? Should we be initializing the 3D template here?

Rohit-Kakodkar commented:
Would be nice to have an example of this. With some plots

lsawade committed:
Merge pull request #666 from PrincetonUniversity/issue-658

Issue 658 -- Implement colors for plotting of EM and Poroelastic domains.

lsawade committed:
Forgot to commit the precommit

lsawade committed:
Updated capitalization of tests/unit-tests/IO -> tests/unit-tests/io

lsawade committed:
Implement new error check

lsawade committed:
Implement new error check

Rohit-Kakodkar committed:
Poroelastic example

Rohit-Kakodkar committed:
Minor fixes

lsawade committed:
Forgot to update the read_sources test

Rohit-Kakodkar committed:
Fixes bug related to property assignment at GLL points

- [ ] Updated the properties assignment
- [ ] Updated constructors for assembly and properties

Rohit-Kakodkar committed:
Merge pull request #670 from PrincetonUniversity/issue-669

Poroelastic example

Commit:
Fixed fetchcontent argument order bug on linux and added target include hdf5 to compute target

icui committed:
Add a new style group

icui committed:
Temporarilty remove group all to avoid installing unused packages.

icui committed:
Update pyproject.toml to specify strict package versions.

arja2 committed:
Merge branch 'PrincetonUniversity:PE' into PE

icui committed:
Use strict version number for packages in pyproject.toml

cmorency1 committed:
modified element hpp

icui committed:
Define strict version number for packages in pyproject.toml

cmorency1 committed:
modified element hpp and cpp

Rohit-Kakodkar committed:
Merge pull request #680 from cmorency1/issue-604

Issue 604 - implemented element type for poroelastic

cmorency1 committed:
modified domain kernels for poroelastic

cmorency1 committed:
created new kernel file for poroelastic

cmorency1 committed:
Edited kernels for poroelastic

Rohit-Kakodkar committed:
Merge pull request #681 from cmorency1/issue-599

Issue 599 - implemented kernels for poroelastic

Commit:
added properties_container

Rohit-Kakodkar committed:
Merge pull request #682 from arja2/issue-602

Issue-602 Implemented properties_container

cmorency1 committed:
Modified wavefield files and stacey boundary container for poroelastic

lsawade committed:
Merge pull request #683 from cmorency1/issue-598

cmorency1 committed:
Modified receiver files for poroelastic

cmorency1 committed:
Modified source files for poroelastic

cmorency1 committed:
Modified data container for storage for poroelastic

lsawade committed:
Merge pull request #648 from PrincetonUniversity/issue-647

Issue 647 - Updating the Github Workflow Doc, Pre-Commit, Change namspace IO -> io

lsawade committed:
Merge pull request #679 from PrincetonUniversity/issue-678

Issue 678 - strict package versions for pyproject.toml

lsawade committed:
Merge pull request #673 from PrincetonUniversity/issue-672

Fixes bug related to property assignment at GLL points

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-566

lsawade committed:
Update the workflow

Rohit-Kakodkar committed:
Updated properties tests with correct PE properties

Rohit-Kakodkar committed:
Merge pull request #686 from cmorency1/issue-603

Issue 603 - Modified data container for storage for poroelastic

Rohit-Kakodkar committed:
Merge pull request #684 from cmorency1/issue-600

Issue 600 - Modified receiver files for poroelastic

Rohit-Kakodkar committed:
Merge pull request #668 from PrincetonUniversity/issue-566

Check timescheme updates all DOFs and elements

Rohit-Kakodkar committed:
Merge pull request #687 from lsawade/PE

Update the github workflow to test upon pull request

Rohit-Kakodkar committed:
Merge pull request #688 from PrincetonUniversity/fixed-PE-properties-tests

Updated properties tests with correct PE properties

Rohit-Kakodkar committed:
Merge branch 'devel' into PE

cmorency1 committed:
Modified the tes files kernels and properties for poroelastic

Rohit-Kakodkar committed:
Merge pull request #691 from cmorency1/issue-603-test

issue 603-test - Modified the tes files kernels and properties for poroelastic

cmorency1 committed:
Modified and added files for mass matrix for poroelastic

Rohit-Kakodkar committed:
Merge pull request #692 from cmorency1/issue-617

issue 617 - Modified and added files for mass matrix for poroelastic

cmorency1 committed:
Modified and created files for wavefields for poroelastic

cmorency1 committed:
Modified and created files for source for poroelastic

cmorency1 committed:
Saved progress on stress implementation for poroelastic

cmorency1 committed:
Edited src compute source file for poroelastic

cmorency1 committed:
Modified timescheme for poroelastic

Rohit-Kakodkar committed:
Merge pull request #695 from cmorency1/issue-613

issue 613 - Modified and created files for wavefields for poroelastic

Rohit-Kakodkar committed:
Merge pull request #696 from cmorency1/issue-614

issue 614 - Modified and created files for source for poroelastic

Rohit-Kakodkar committed:
Merge pull request #698 from cmorency1/issue-612

issue 612 - Modified timescheme for poroelastic

cmorency1 committed:
Modified stress for poroelastic

cmorency1 committed:
Merge branch 'PE' into issue-616

Rohit-Kakodkar committed:
Merge pull request #697 from cmorency1/issue-616

issue 616 - Saved progress on stress implementation for poroelastic

cmorency1 committed:
Ran displacement test!!

Rohit-Kakodkar committed:
Fixed damping term implementation

lsawade committed:
Merge pull request #699 from cmorency1/issue-611

issue 611 - Ran displacement test!!

cmorency1 committed:
Added viscous damping file for poroelastic

lsawade committed:
Updated the signal throwing

cmorency1 committed:
Modifed file for viscous damping poroelastic

lsawade committed:
Updated the core.cpp

lsawade committed:
updated indent

lsawade committed:
Updated the files

Rohit-Kakodkar committed:
Merge pull request #701 from cmorency1/issue-700

issue 700 - Added viscous damping file for poroelastic

Rohit-Kakodkar committed:
Partial implementation of stacey BCs

I noticed that there no contribution to mass matrix from Stacey BC. Waiting for response from Christina that this isn't an oversight.

Rohit-Kakodkar committed:
Fixed formatting issues

Rohit-Kakodkar committed:
Fixed cmake presets

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-663

lsawade committed:
Enforced not VTK_CXX_BUILD=OFF which was disregarded if VTK was found, and implemented working Ctrl-C check

icui committed:
Remove comments from json

lsawade committed:
Updated the TAG MEDIUM_TAG_ELASTIC_SV -> MEDIUM_TAG_ELASTIC_PSV and specfem::element::medium_tag::elastic_sv -> specfem::element::medium_tag::elastic_psv

icui committed:
Fix Kokkos links in documentataion.

lsawade committed:
Changed electromagnetic_sv -> electromagnetic_te

lsawade committed:
Moved check_signal syntax to cpp file

cmorency1 committed:
Modified and created files for frechet derivatives for poroelastic

Rohit-Kakodkar committed:
Merge pull request #716 from PrincetonUniversity/issue-547

Issue 574 - Fixing the execution from Python + VTK_CXX_BUILD=OFF works now even if VTK is found

Rohit-Kakodkar committed:
Merge pull request #718 from PrincetonUniversity/issue-717

Issue 717 - Fix Kokkos links in documentataion.

icui committed:
Add CMakeUserPresets.json to .gitignore

lsawade committed:
Updated the elastic_wave tag p_sv -> psv

Rohit-Kakodkar committed:
Full implementation for stacey BCs

lsawade committed:
Merged

lsawade committed:
Updated all tests to include the call to EM wave

Rohit-Kakodkar committed:
Implemented max reduction for compute properties / kernels

- [ ] Implemented general reduction operator
- [ ] Enabled max reductions
- [ ] Enabled assembly check for PE & stacey elements

lsawade committed:
Updated the is_elastic struct

lsawade committed:
is_elastic and is_electromagnetic implemented for materials

lsawade committed:
Updated kernel_ and properties_containers

lsawade committed:
Final update to the point kernels

lsawade committed:
Merge pull request #719 from PrincetonUniversity/issue-629

Issue 629 - Update the medium_tag elastic_sv -> elastic_psv

Rohit-Kakodkar committed:
Merge pull request #664 from PrincetonUniversity/issue-663

Fix CMake preset for cuda

lsawade committed:
Merge pull request #720 from PrincetonUniversity/issue-674

lsawade committed:
Update

icui committed:
Re-organize macros.

icui committed:
Merge pull request #749 from PrincetonUniversity/issue-748

Updated cmakelists to get git repo

icui committed:
Merge branch 'devel' into issue-752

Rohit-Kakodkar committed:
Merge pull request #721 from cmorency1/issue-615

issue 615 - Modified and created files for frechet derivatives for poroelastic

Rohit-Kakodkar committed:
Merge pull request #745 from PrincetonUniversity/issue-708

Issue 708 - Stacey BC implementation for poroelastic

Rohit-Kakodkar committed:
Implemented register arrays that use std::layouts and std::extents as accessors

Rohit-Kakodkar committed:
Added Kokkos function macros | made operators constexprs

lsawade committed:
Fix for pyproject toml

icui committed:
Rename SV to PSV

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-755

icui committed:
Merge branch 'devel' into PE

icui committed:
Update poroelastic element types.

icui committed:
Merge branch 'PE' of github.com:PrincetonUniversity/SPECFEMPP into PE

icui committed:
Merge branch 'devel' into PE

icui committed:
Update formatting.

icui committed:
Merge pull request #690 from PrincetonUniversity/PE

Poroleastic implementation

icui committed:
Merge pull request #753 from PrincetonUniversity/issue-752

Issue 752 - Re-organize current macros

Rohit-Kakodkar committed:
Fixed loop index shadowing

Rohit-Kakodkar committed:
Merge pull request #756 from PrincetonUniversity/issue-755

Rohit-Kakodkar committed:
Fixed NVIDIA compilation

- [x] Moved reduce member function to standalone function
- [x] Made the function a friend function

Rohit-Kakodkar committed:
Added KOKKOS_LAMBDA to reducer lambda

icui committed:
Add macro FOR_EACH_MATERIAL_SYSTEM

icui committed:
Update capture type.

icui committed:
Implement new macro for element_types and value_containers

icui committed:
Update capture macro

icui committed:
Update macro for io and receiver.

icui committed:
Update macro for materials

icui committed:
Add underscore to macro reference variable

icui committed:
Fix TE tag

icui committed:
Update macro for mesh

icui committed:
Update type macro syntax

icui committed:
Update macro for instantiation

icui committed:
Update material instantiation.

icui committed:
Update boost vmd import.

icui committed:
Rename WHERE2 to FROM

icui committed:
Update cmake for vmd

icui committed:
Update BOOST_LIBS

icui committed:
Update macro for tests

Rohit-Kakodkar committed:
Upated field derivatives and stress tensor layout

- [x] Updates `VectorPointViewType` with components as first extent and dimension as second extent
- [x] Updates `point::field_derivatives`
- [x] Updates `point::stress`
- [x] Updates the `gradient` operator
- [x] Updates `stress`, `wavefield`, & `frechet_derivatives` computations for different media

Rohit-Kakodkar committed:
Added Kokkos inline function macro

icui committed:
Update material operator[]

Rohit-Kakodkar committed:
Updated chunk vector view layouts

- [x] Updates chunk vector view layouts
- [x] Update the storage layout for chunk stress integrands (F)

icui committed:
Update comment for macro FROM

icui committed:
Change variant to any for material tests.

icui committed:
Remove macro _CALL_FOR_ALL_MATERIAL_SYSTEMS

icui committed:
Rename macro FROM to IN

icui committed:
Rename macro FROM to IN

icui committed:
Merge branch 'issue-704' into issue-707

lsawade committed:
Updated compilation and test workflows

lsawade committed:
Updated the comment in the unit test

icui committed:
Generalize some macros in material_systems.

icui committed:
Generalize material system macros.

icui committed:
Rename FOR_EACH_MATERIAL_SYSTEM and IN

icui committed:
Partially replace macros for ELEMENT_TYPES

Rohit-Kakodkar committed:
Fixed Intel CI after system update

Skylake nodes were removed from the SLURM queue in the last system downtime. Update intel CI slurm constraint to `intel|cascade`.

Rohit-Kakodkar committed:
Fixed GNU tests constraint

lsawade committed:
Merge branch 'CI-fix' into workflow-fix

Rohit-Kakodkar committed:
Merge pull request #770 from PrincetonUniversity/CI-fix

Fixed Intel CI after system update

icui committed:
Update semicolon in macro

Rohit-Kakodkar committed:
Merge pull request #767 from PrincetonUniversity/workflow-fix

Updated compilation and test workflows

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-693-1

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-693

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-704

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-746

Rohit-Kakodkar committed:
Updated compilers available on della9

Rohit-Kakodkar committed:
Merge pull request #775 from PrincetonUniversity/rh9-update-fix

Updated compilers available on della9

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-693-1

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-693

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-746

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-704

Rohit-Kakodkar committed:
Merge pull request #754 from PrincetonUniversity/issue-746

icui committed:
Merge branch 'devel' into issue-704

icui committed:
Merge pull request #762 from PrincetonUniversity/issue-704

Issue 704, 705, 706 - new macros for code block, variable declaration and template instantiation

icui committed:
Remove explicit instantion of boundary conditions.

icui committed:
Merge branch 'devel' into issue-707

icui committed:
Update to new macro for ELEMENT_TYPES.

icui committed:
Fix test_load.

icui committed:
Merge branch 'issue-778' into issue-707

icui committed:
Update source test and remove old macros for ELEMENT_TYPES.

icui committed:
Add support for accessing class properties from macros

icui committed:
Update to new macro for medium_tags

icui committed:
Clean up unused macros.

icui committed:
Rename macro FOR_EACH to FOR_EACH_IN_PRODUCT.

icui committed:
Change macro syntax from TAG_XXX, ... to TAG(XXX, ...

icui committed:
Re-organize source test.

icui committed:
Update macro for medium io

icui committed:
Disable check_load for future fix.

icui committed:
Update check_load.

lsawade committed:
Updated the parameter docs

icui committed:
Fix test for assembly sources on GPU.

icui committed:
Merge branch 'issue-778' into issue-780

icui committed:
Add [[maybe_unused]] for macros to avoid compiler warning.

icui committed:
Update parameter of periodic_task.

icui committed:
Rename variables.

lsawade committed:
Created a assembly print function.

icui committed:
Make wavefield reader / writer periodic tasks.

icui committed:
Update jpg file name for plotter.

icui committed:
Fix capture error for cuda 11.8

icui committed:
Merge pull request #779 from PrincetonUniversity/issue-778

Issue 778 - fix function check_load in assembly source test

icui committed:
Merge branch 'issue-778' into issue-780

icui committed:
Update cmake for string utility.

icui committed:
Merge branch 'devel' into issue-780

icui committed:
Cmake fix.

Rohit-Kakodkar committed:
Merge pull request #785 from PrincetonUniversity/issue-784

Issue 784 - Updated the parameter doc & added print function for assembly

icui committed:
Merge pull request #787 from PrincetonUniversity/issue-783

Issue 783 - Add option to write wavefield every N timesteps

icui committed:
GPU fix for compute_sources.

icui committed:
Merge branch 'devel' into issue-780

icui committed:
Update macro.

Rohit-Kakodkar committed:
Updates to mesher parameters to run CUBIT

- [x] Updates the fortran mesher to mesh NGNOD=9
- [x] Fixes fortran mesher bug to read external absorbing boundary files
- [x] Updates SPECFEM++ mesh reader to NGNOD=9 meshes

icui committed:
Use macro for wavefield reader / writer.

Rohit-Kakodkar committed:
Merge pull request #792 from PrincetonUniversity/issue-791

Updates to mesher parameters to run CUBIT

icui committed:
Bug fix for undefined writer
time interval

icui committed:
Merge branch 'devel' of github.com:PrincetonUniversity/SPECFEMPP into devel

icui committed:
Merge branch 'devel' into issue-793

icui committed:
Wavefield reader bug fix.

icui committed:
Merge pull request #782 from PrincetonUniversity/issue-780

Issue 780 - Update to new macro for MEDIUM_TAGS and ELEMENT_TYPES

icui committed:
Re-organize wavefield reader / writer.

icui committed:
Write wavefield data of different time steps into a single file.

icui committed:
Merge branch 'devel' into issue-793

icui committed:
Make io::reader / writer pure virtual.

lsawade committed:
Added damping_force flag to medium enumerations.

lsawade committed:
Added dispatch for damping computation so that not every dimension/medium/property combination needs one.

lsawade committed:
Updated the setup to move the attributes class using default attributes.

icui committed:
Fix compilation error.

icui committed:
Add string utilities to relax configuration string

icui committed:
Update tests.

lsawade committed:
Removed the default attributes

lsawade committed:
Renamed the file

lsawade committed:
Updated the damping force computation execution removing the activated dispatch.

lsawade committed:
Removed the seismogram type and replacede it with wavefield type

lsawade committed:
Added always false templated type for better static_assert error for required specialization.

icui committed:
Change flush scope to local.

icui committed:
Merge pull request #796 from PrincetonUniversity/issue-793

Issue 793 - Write wavefield data of different time steps into a single file.

lsawade committed:
Merge branch 'devel' into issue-797

lsawade committed:
Made dimension/components variables instead of functions, updated impl_compute_damping_call

lsawade committed:
things work. but still feedback

lsawade committed:
(hopefully) final implementation

lsawade committed:
Update compute damping force implementation

icui committed:
Use macros for config string test functions.

lsawade committed:
Merge pull request #798 from PrincetonUniversity/issue-797

Issue 797 - Implements compile-time conditional execution of "extra physics" [poroelastic damping]

icui committed:
Optimize compilation for macros

icui committed:
Simplify macros for arrays.

icui committed:
Bug fix.

icui committed:
Update macro WRITE_TAGS

lsawade committed:
Merge pull request #799 from PrincetonUniversity/issue-773

Removed the seismogram type and replaced it with wavefield type

icui committed:
Rename DimensionType, MediumType to DimensionTag, MediumTag

Rohit-Kakodkar committed:
Added KOKKOS_FUNCTION decorator

Rohit-Kakodkar committed:
Merge pull request #764 from PrincetonUniversity/issue-693-1

Updated chunk vector view layouts

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-693

Rohit-Kakodkar committed:
Merge pull request #763 from PrincetonUniversity/issue-693

Upated field derivatives and stress tensor layout

icui committed:
Merge pull request #806 from PrincetonUniversity/issue-805

Issue 805 - Optimize compilation for macros

icui committed:
Merge branch 'devel' into issue-790

icui committed:
Merge pull request #803 from PrincetonUniversity/issue-790

Issue 790 - Add string utilities to relax configuration string

icui committed:
Merge pull request #820 from PrincetonUniversity/issue-810

Issue 810 - Rename DimensionType, MediumType to DimensionTag, MediumTag

int-ptr-ptr committed:
Merge branch 'devel' into subdivisions

int-ptr-ptr commented:
Reviving this PR now that generals are over. Will merging devel into [PrincetonUniversity:issue-516](https://github.com/PrincetonUniversity/SPECFEMPP/tree/issue-516) clean up the [Files changed](https://github.com/PrincetonUniversity/SPECFEMPP/pull/474/files) tab of this PR?

Rohit-Kakodkar commented:
I think the best way forward would be to close this PR as a stale PR. It would be a lot of unnecessary effort to resolve branch `516`, which it seems like you have already done for your branch. Why dont you issue a new PR into `devel` and we go from there.
